packages/network-controller/src/NetworkController.test.ts

```diff
@@ -1,278 +1,4156 @@
+
+  controller.providerConfig = buildProviderConfig();
+  await waitForResult(true, () => controller.provider !== undefined);
+  assert(controller.provider);
+
+  if (stubLookupNetworkWhileSetting) {
+    lookupNetworkMock.mockRestore();
+  }
+  if (stubGetEIP1559CompatibilityWhileSetting) {
+    lookupGetEIP1559CompatibilityMock.mockRestore();
+  }
+
+  return controller.provider;
+}
+
+/**
+ * Waits for a controller event to be emitted. If given a function, that
+ * function will be called first before waiting.
+ *
+ * @param messenger - The messenger suited for NetworkController.
+ * @param eventType - The type of event you want to wait for.
+ * @param rest - Either an options bag and a function, or a function, or
+ * nothing. The options bag may contain `count`, the number of events
+ * (default: 1); `filterEventPayload`, a function to filter the events; and
+ * `wait`, the amount of time to wait before assuming all events have occurred.
+ * The function will be called before the promise resolves.
+ * @returns A promise that resolves to the list of event payloads (optionally
+ * filtered) when the specified number of them have occurred.
+ */
+async function waitForPublishedEvents<E extends NetworkControllerEvents>(
+  messenger: NetworkControllerMessenger,
+  eventType: E['type'],
+  ...rest:
+    | [
+        {
+          count?: number;
+          filterEventPayload?: (payload: E['payload']) => boolean;
+          wait?: number;
+        },
+        () => void | Promise<void>,
+      ]
+    | [() => void | Promise<void>]
+    | []
+): Promise<E['payload'][]> {
+  let options: {
+    count?: number;
+    filterEventPayload?: (payload: E['payload']) => boolean;
+    wait?: number;
+  } = {};
+  let fn: () => void | Promise<void> = () => {
+    // do nothing
+  };
+  if (rest.length === 2) {
+    options = rest[0];
+    fn = rest[1];
+  } else if (rest.length === 1) {
+    fn = rest[0];
+  }
+  const {
+    count = 1,
+    filterEventPayload = () => true,
+    wait: timeBeforeAssumingNoMoreEvents = 150,
+  } = options;
+
+  const promiseForEvents = new Promise<E['payload'][]>((resolve, reject) => {
+    // We need to declare this variable first, then assign it later, so that
+    // ESLint won't complain that resetTimer is referring to this variable
+    // before it's declared. And we need to use let so that we can assign it
+    // below.
+    /* eslint-disable-next-line prefer-const */
+    let eventListener: (...args: E['payload']) => void;
+    let timer: NodeJS.Timeout | undefined;
+    const allEventPayloads: E['payload'][] = [];
+    const interestingEventPayloads: E['payload'][] = [];
+    let alreadyEnded = false;
+
+    const end = () => {
+      if (!alreadyEnded) {
+        messenger.unsubscribe(eventType.toString(), eventListener);
+        if (interestingEventPayloads.length === count) {
+          resolve(interestingEventPayloads);
+        } else {
+          // Using a string instead of an Error leads to better backtraces.
```

Comment:
In what manner does this improve stack traces? This seems unconventional.

Reply:
Right, good question. So if I use an Error object here then this is the full error I get:

```
    Expected to receive 1 NetworkController:providerConfigChange event(s), but received 0 after 150ms.

    All payloads:

    []

      4019 |             /* eslint-disable-next-line prefer-promise-reject-errors */
      4020 |             reject(
    > 4021 |               new Error(
           |               ^
      4022 |                 `Expected to receive ${expectedNumberOfEvents} ${eventType} event(s), but received ${
      4023 |                   interestingEventPayloads.length
      4024 |                 } after ${timeBeforeAssumingNoMoreEvents}ms.\n\nAll payloads:\n\n${inspect(

      at end (src/NetworkController.test.ts:4021:15)
      at src/NetworkController.test.ts:4044:11

```

The stack trace refers to the line that is shown as well as the call to `end()` a little further down, but then it stops there. Of course Jest prints the name of the test that failed, but given how many tests there are, that is a bit difficult to find.

If I don't use an Error object, however, then I get:

```
    Failed: "Expected to receive 1 NetworkController:providerConfigChange event(s), but received 0 after 150ms.

    All payloads:

    []"

      2720 |
      2721 |         describe('if the network version could be retrieved during the call to setRpcTarget', () => {
    > 2722 |           it('does not retrieve the network version again', async () => {
           |           ^
      2723 |             const messenger = buildMessenger();
      2724 |             await withController({ messenger }, async ({ controller }) => {
      2725 |               const fakeMetamaskProvider = buildFakeMetamaskProvider([

      at Env.it (../../node_modules/jest-jasmine2/build/jasmineAsyncInstall.js:95:24)
      at Suite.<anonymous> (src/NetworkController.test.ts:2722:11)
      at Suite.<anonymous> (src/NetworkController.test.ts:2721:9)
```

Now we can see that the test that failed starts at line 2722.

It appears that this is due to our use of Node 14. If I switch to Node 16, then I get better stack traces when using an Error object.

Reply:
Huh. Interesting finding! I wouldn't have expected that. I hate to see us using anything but an `Error` for errors (in the past this has led to super confusing debugging experiences, incompatibilities with tools, etc). but I guess it makes sense here at least until we update to Node.js v16.